src/internals/parser/Parser.ts

```diff
@@ -0,0 +1,742 @@
+    ...infer Tail extends ParserFn[]
+  ]
+    ? Call<Head, Input> extends infer A
+      ? A extends Ok
+        ? A
+        : ChoiceImpl<Self, Tail, Input, [...ErrorAcc, ErrMsg<A>]>
+      : never
+    : Err<Self, Input, ErrorAcc>;
+
+  /**
+   * Parser that tries to match the input with one of the given parsers.
+   * @description The parsers are tried in the order they are given.
+   * @param Parsers - the parsers to try
+   * @returns an Ok type if one of the parsers matches or an error with all the errors of the parsers
+   *
+   * @example
+   * ```ts
+   * type T0 = Call<Choice<[
+   * //   ^? type T0 = Ok<{ result: "a"; input: "" }>
+   *   Literal<"a">,
+   *   Literal<"b">,
+   * ]>, "a">;
+   * type T1 = Call<Choice<[
+   * //   ^? type T1 = Ok<{ result: "b"; input: "" }>
+   *  Literal<"a">,
+   *  Literal<"b">,
+   * ]>, "b">;
+   * type T2 = Call<Choice<[
+   * //   ^? type T2 = Error<{ message: "Expected 'choice(literal('a'),literal('b'))' - Received 'c'"; cause: "Expected 'literal('a')' - Received 'c' | Expected 'literal('b')' - Received 'c'";}>
+   * Literal<"a">,
+   * Literal<"b">,
+   * ]>, "c">;
+   * ```
+   */
+  export interface Choice<Parsers extends ParserFn[]> extends ParserFn {
+    name: "choice";
+    params: Parsers;
+    return: this["arg0"] extends infer Input extends string
+      ? ChoiceImpl<this, Parsers, Input>
+      : InputError<this["arg0"]>;
+  }
+
+  /**
+   * Parser that tries to match the input with one of the two given parsers.
+   * @description The parsers are tried in the order they are given.
+   * @param Parser1 - the first parser to try
+   * @param Parser2 - the second parser to try
+   * @returns an Ok type if one of the parsers matches or an error with all the errors of the parsers
+   *
+   * @example
+   * ```ts
+   * type T0 = Call<Or<Literal<"a">, Literal<"b">>, "a">;
+   * //   ^? type T0 = Ok<{ result: "a"; input: "" }>
+   * ```
+   */
+  export type Or<Parser1 extends ParserFn, Parser2 extends ParserFn> = Choice<
+    [Parser1, Parser2]
+  >;
+
+  /**
+   * Parser that optionally matches the input with the given parser.
+   * @description If the parser matches it will return the result of the parser.
+   * If the parser doesn't match it will return an empty array.
+   * @param Parser - the parser to match
+   * @returns an Ok type if the parser matches and an empty array or the error of the parser
+   *
+   * @example
+   * ```ts
+   * type T0 = Call<Optional<Literal<"a">>, "a">;
+   * //   ^? type T0 = Ok<{ result: ["a"]; input: "" }>
+   * type T1 = Call<Optional<Literal<"a">>, "b">;
+   * //   ^? type T1 = Ok<{ result: []; input: "b" }>
+   * ```
+   */
+  export interface Optional<Parser extends ParserFn> extends ParserFn {
+    name: "optional";
+    params: [Parser];
+    return: this["arg0"] extends infer Input extends string
+      ? Call<Parser, Input> extends infer A
+        ? A extends Ok
+          ? A
+          : Ok<{ result: []; input: Input }>
```

Comment:
This accidentally says `Ok<{ result: []; input: Input }>` instead of `Ok<[], Input>`

Reply:
Thanks you, yes,  just fixed it, but i obviously need to add some unit tests.

My main concern about this PR is the fact that `Recursive parsers` don't work and this might make the whole thing unusable.

So all feadback and ideas on how to fix that is welcomed

Reply:
Looks like you got it to work, awesome!